Making jenkins_plugin FIPS 140-2 compliant

jesstruck opened the pull request:
##### SUMMARY

Fixes Issue #34304.

The hashlib.md5(...) function is disabled on systems running in FIPS 140-2 mode. To accommodate that, this commit just switches to a hash function (hashlib.sha1(...)) that is allowed by FIPS 140-2.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`jenkins_plugin`

##### ADDITIONAL INFORMATION
This is a second try of getting [34349](https://github.com/ansible/ansible/pull/34349) accepted and merged in, I have simply reimplemented @karlmdavis solution in the previous mentioned pull request, together with the suggested changes he got.

jesstruck committed:
replace the use of MD5 to use sha1 instead to make the module support FIPS 140-2 support

jesstruck committed:
replace the use of MD5 to use sha1 instead to make the module support FIPS 140-2 support

jesstruck committed:
fixing copy/paste error

jesstruck committed:
fixing copy/paste error

jesstruck committed:
fixing syntaxt error based on shipable

jesstruck committed:
fixing syntaxt error based on shipable

jtyr reviewed (approved):
LGTM

gundalow commented:
I'm not that familiar with FIPS, though I wonder if we should add a "code smell" CI Test, so that we don't allow other code to use `hashlib.md5`, @abadger1999 @mattclay what are your thoughts?

mattclay commented on lib/ansible/modules/web_infrastructure/jenkins_plugin.py:
```diff
@@ -427,11 +427,13 @@ def install(self):
+                self.module.fail_json(
+                    msg="Cannot calculate SHA1 of the old plugin.",
+                    details="SHA1 calculation failed with: {0}".format(err)
+                )
```
Why the additional changes beyond switching the hash algorithm?

This appears to change the behavior of the module. Before a non-existent `plugin_file` would result in the hash being `None`, but now it results in a module failure.

jesstruck replied:
Correct, this comes directly from the previous PR, (Se original msg). I have no problem removing it.

mattclay commented on lib/ansible/modules/web_infrastructure/jenkins_plugin.py:
```diff
@@ -448,31 +450,39 @@ def install(self):
                     self.params['updates_expiration'] == 0 or
                     self.params['version'] not in [None, 'latest'] or
-                    md5sum_old is None):
+                    sha1sum_old is None):
```
There doesn't appear to be any way for `sha1sum_old` to be `None`.

mattclay commented on lib/ansible/modules/web_infrastructure/jenkins_plugin.py:
```diff
@@ -448,31 +450,39 @@ def install(self):
 
                 # Write downloaded plugin into file if checksums don't match
-                if md5sum_old is None:
+                if sha1sum_old is None:
```
There doesn't appear to be any way for `sha1sum_old` to be `None`.

jesstruck replied:
this seems like it should have been `sha1_old` and not `sha1sum_old`

mattclay commented on lib/ansible/modules/web_infrastructure/jenkins_plugin.py:
```diff
@@ -448,31 +450,39 @@ def install(self):
+                            msg="Cannot calculate SHA1 of the downloaded plugin.",
+                            details="SHA1 calculation failed with: {0}".format(err)
+                        )
+                    sha1sum_new = base64.b64encode(sha1_new.digest())
```
Under what condition would this exception handler even run?

jesstruck replied:
I would guess that it could only throw an exception, in the scenario, where there is not access to the plugin_url, and thereby the `data` will be null, you could very well argue such a check should b around `r.read()` and not around the later usage. Let me implement all the suggestion and come back with a new implementation. I will also merge master in

dagwieers committed:
Fix various sonarcloud issues

This fixes various reported bugs through sonarcloud at:
https://sonarcloud.io/project/issues?id=Rodney-Reis_ansible&resolved=false&types=BUG

dagwieers committed:
Assorted pylint fixes

Commit:
recoding after the last code review

Commit:
recoding after the last code review

dagwieers committed:
crypto: Fix known issues in modules (#52302)

* crypto: Fix known issues in modules

This fixes a few issues reported by 'validate-modules'.

* Fix whitespace

Commit:
merge devel branch into this PR

dagwieers committed:
Fix module issues (#52209)

dagwieers committed:
doc_fragments: Clean up parameter types (cloud) (#52177)

* module_utils: Clean up parameter types (cloud)

This PR includes:

- Parameter types added
- Copyright format fixes
- Short license statement
- Description fixes (only for a few files, then I stopped :-))

* More cloud stuff

* Fix reported issue

* Fix quotes

* Use correct type

* Fix quotes

* Fix quotes

* Fix quotes

* Fix typo

* Fix boolean

migumun committed:
Fortinet's FortiOS report theme (#52298)

migumun committed:
Fortinet's FortiOS report style (#52296)

migumun committed:
Fortinet's FortiOS report setting (#52295)

migumun committed:
Fortinet's FortiOS report layout (#52294)

migumun committed:
Fortinet's FortiOS log webtrends setting (#52273)

migumun committed:
Fortinet's FortiOS log threat weight (#52271)

migumun committed:
Fortinet's FortiOS log syslogd setting (#52270)

migumun committed:
Fortinet's FortiOS log syslogd override setting (#52269)

migumun committed:
Fortinet's FortiOS log syslogd override filter (#52268)

* Fortinet's FortiOS log syslogd override filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log syslogd filter (#52267)

* Fortinet's FortiOS log syslogd filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log syslogd4 setting (#52266)

migumun committed:
Fortinet's FortiOS log syslogd4 filter (#52265)

* Fortinet's FortiOS log syslogd4 filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log syslogd3 setting (#52264)

migumun committed:
Fortinet's FortiOS log syslogd2 setting (#52261)

migumun committed:
Fortinet's FortiOS log fortianalyzer override filter (#52163)

* Fortinet's FortiOS log fortianalyzer override filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log fortianalyzer filter (#52162)

* Fortinet's FortiOS log fortianalyzer filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log fortianalyzer2 filter (#52157)

* Fortinet's FortiOS log fortianalyzer2 filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log eventfilter (#52151)

migumun committed:
Fortinet's FortiOS log disk setting (#52150)

migumun committed:
Fortinet's FortiOS log disk filter (#52149)

* Fortinet's FortiOS log disk filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log custom field (#52148)

migumun committed:
Fortinet's FortiOS ips rule (#52121)

* Fortinet's FortiOS ips rule

* Avoid null choices

* Avoid null choices

migumun committed:
Fortinet's FortiOS log fortiguard filter (#52167)

* Fortinet's FortiOS log fortiguard filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS system interface (#52247)

migumun committed:
Fortinet's FortiOS log fortiguard override filter (#52168)

* Fortinet's FortiOS log fortiguard override filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log memory filter (#52250)

* Fortinet's FortiOS log memory filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log memory setting (#52255)

migumun committed:
Fortinet's FortiOS log memory global setting (#52251)

migumun committed:
Fortinet's FortiOS log null device filter (#52256)

* Fortinet's FortiOS log null device filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS log null device setting (#52257)

migumun committed:
Fortinet's FortiOS log setting (#52258)

migumun committed:
Fortinet's FortiOS log syslogd2 filter (#52259)

* Fortinet's FortiOS log syslogd2 filter

* Avoid null choices

migumun committed:
Fortinet's FortiOS Firewall Schedule OneTime (#52076)

migumun committed:
Fortinet's FortiOS fortios_firewall_schedule_recurring (#52078)

Andersson007 committed:
Module postgresql_idx: added ci tests, new params, returned values, code refactoring (#52230)

* postgresql_idx: ci tests, refactoring, return values

* postgresql_idx: ci tests, new params, return values

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* postgresql_idx: ci tests, fix

* New module postgresql_table - fix tests

* New module postgresql_table - fix tests

* New module postgresql_table - fix tests

* New module postgresql_table - fix state choices order

dsoper2 committed:
Cisco Intersight module_utils and intersight_facts module (#51309)

* Cisco Intersight module_utils and intersight_facts module

* Add RETURN information and fix pylint, import, and pep8 issues.

* Review updates for specifying type of params/returns and not polluting ansible_facts.

* BSD one line license, validate_certs used, urls.fetch_urls replaces requests

timorunge committed:
module_utils/mysql: Fixing unexpected keyword argument 'cursorclass' error after migratio… (#47809)

* Fixing unexpected keyword argument 'cursorclass' error after migration from MySQLdb to PyMySQL

* Adoptions for mysql.py as suggested by felixfontein.

* Adding changelog fragment.

dagwieers committed:
Sanity fixes to parameter types (#52325)

Akasurde committed:
distribution: Add support for Cumulus Linux (#52309)

* Refactored code
* Added support for Cumulus Linux 2.5.4
* Added support for Cumulus Linux 3.7.3
* Test added

Fixes: #29969

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

mattclay committed:
Fix sanity test failures (pep8).

mkrizek committed:
Native Jinja2: raise undefined error immediately (#52237)

Fixes #52158

daurrutia committed:
doc: remove grammatical glitch (#52336)

mattclay committed:
Unfreeze six in test constraints.

This avoids re-installation during integration test runs.

Pinning this requirement shouldn't be needed for consistent test
results when running pylint.

dsummersl committed:
cron - adds deprecation warnings (#37355). (#52322)

Adds deprecation warnings for both 'name', and 'reboot' option (although
not related to this change, it has been deprecated for quite a while)

mattclay committed:
Fix tests to use ANSIBLE_TEST_PYTHON_INTERPRETER.

dagwieers committed:
Fix one issue (#52370)

webknjaz committed:
Bugfix/ fix mocker patch in tests (#52372)

* 🐛 Fix invalid os.stat mock in tests

* 🐛 Fix leaking mock patch in tests

Closes #52347

Akasurde committed:
validate-modules: Documentation bool (#50085)

* validate-modules: Documentation bool

This check allows to catch cases where type of argument is different than documentation does.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

* Wrong comparison for 'str'

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

* Add ignore.txt

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

* Fix logic and clean up ignore.txt

mattclay committed:
Update pytest constraint for Python 2.7.

saranyasridharan committed:
Adding custom module to get PID of the process (#50896)

mattclay committed:
Update the default container to version 1.5.2.

ci_complete

mattclay committed:
Allow setting resource.RLIMIT_NOFILE in modules (#51989)

richlv committed:
clarify how users are specified (#52260)

##### SUMMARY
Clarify that usernames - which Zabbix API documentation calls "aliases" - are to be specified for recipients

##### ISSUE TYPE
- Docs Pull Request

+label: docsite_pr

felixfontein committed:
docker_* modules: remove from validate-modules ignore list (#52402)

* Remove docker_* modules from validate-modules ignore list.

* Adjust types of cacert_path, cert_path and key_path.

* Fix type of update_failure_action in docker_swarm_service.

WojciechowskiPiotr committed:
Integration tests for module docker_swarm_facts (#52208)

* * docker_swarm_facts integration tests

* * docker_swarm_facts integration tests

* * docker_swarm_facts integration tests

* Removing commented lines

* Extended and updated integration tests for docker_swarm_facts module

* Cosmetic integtration tests playbook changes

* Cosmetic integtration tests playbook changes from PR comments

vfauth committed:
etcd3: Support authentication (#44383)

* etcd3: Support authentication

* etcd3: Comment argument check

* etcd3: Fixed wrong var names

* etcd3: Added in 2.8 instead of 2.7

Andy-ch committed:
Correct minor typo in filesystem module docs (#52413)

Changed 'filesytem' to 'filesystem'

Akasurde committed:
pacman: Check for package list before installing (#52246)

Fixes: #45638

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

migumun committed:
Fortinet's FortiOS log syslogd3 filter (#52262)

* Fortinet's FortiOS log syslogd3 filter

* Avoid Null Choices

WojciechowskiPiotr committed:
Integration tests for docker_node_facts (#52423)

felixfontein committed:
docker_common: improve image finding methods (#44692)

* Work around problem with Docker daemon that sometimes won't find image if prefixed with docker.io repo name.

* When tring library/xxx, docker-py also doesn't sometimes find the image.

* Add changelog.

anis016 committed:
Fixed typos (#52426)

Commit:
fix version import in common (#52005)

Akasurde committed:
Distribution: Add support for Linux Mint 18.3 (#52224)

* Distribution: Add support for Linux Mint 18.3
* Review comments

Fixes: #52210

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

jborean93 committed:
opensuse - install password-store from specific repo (#52439)

WojciechowskiPiotr committed:
Fix typo in key name (#52432)

shinmog committed:
Bug fixes and enhancements for panos connection (#52306)

Commit:
azure_rm_cosmosdbaccount - fix for virtual network rules (#52416)

jborean93 committed:
zypper: fix tests to use new URL for OpenSUSE 15.0 (#52453)

jborean93 committed:
zypper_repository: fix return check on newer OpenSUSE versions (#52457)

yuwzho committed:
fix non location error (#52312)

Commit:
Fixing azure_rm_postgresqldatabase force_update parameter (#52388)

jborean93 committed:
alternatives - fix OpenSUSE dir on newer distros (#52458)

optionsome committed:
Fix az command to get object id (#52277)

Commit:
fixing force_update in azure_rm_mysqldatabase (#52389)

Commit:
adding postgresqlconfiguration facts (#45071)

yuwzho committed:
modified storageaccount fact return curated result with connection string (#49702)

Commit:
Merge branch 'jenkins_plugins_sha1' of github.com:jesstruck/ansible into jenkins_plugins_sha1

Commit:
removing whitespaces from empty line

gundalow commented:
Please reopen once you've rebased this PR and resolved the conflicts https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html
